File: coding/src/com/algos/li07_stack/revised/L225_ImplementStackUsing2Queues.java
package  com.algos.li07_stack.revised;

import java.util.LinkedList;
import java.util.Queue;

public class L225_ImplementStackUsing2Queues {

	public static void main(String[] args) {
		L225_ImplementStackUsing2Queues obj = new L225_ImplementStackUsing2Queues();
		
		obj.push(1);
		obj.push(2);
		System.out.println(obj.top());
		System.out.println(obj.pop());
		System.out.println(obj.empty());
	}
	
    // Problem: https://leetcode.com/problems/implement-stack-using-queues/
    // Idea: Use two queues. Bofore inserting move all the elements to q2 and push the new element q1 and move back all the elements to q1.
	Queue<Integer> queue1;
    Queue<Integer> queue2;
    public L225_ImplementStackUsing2Queues() {
        queue1 = new LinkedList<>();
        queue2 = new LinkedList<>();
    } 
    
    public void push(int x) {
        while(!queue1.isEmpty())
            queue2.add(queue1.remove());
        queue1.add(x);
        while(!queue2.isEmpty())
            queue1.add(queue2.remove());
    }
    
    public int pop() {
        return queue1.remove();
    }
    
    public int top() {
        return queue1.peek();
    }
    
    public boolean empty() {
        return queue1.isEmpty();
    }

}

File: dsa/patterns/07_stacks/python/l225_implement_stack_using2_queues.py
"""
Problem: https://leetcode.com/problems/implement-stack-using-queues/
Idea:
Use two queues.
Before inserting a new element, move all elements from queue1 to queue2.
Push the new element into queue1.
Move everything back from queue2 to queue1 so that the newest element
always stays at the front of queue1.
"""

from collections import deque

class MyStack:

    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x: int) -> None:
        # Move all elements to q2
        while self.q1:
            self.q2.append(self.q1.popleft())

        # Push new element to q1
        self.q1.append(x)

        # Move everything back to q1
        while self.q2:
            self.q1.append(self.q2.popleft())

    def pop(self) -> int:
        return self.q1.popleft()

    def top(self) -> int:
        return self.q1[0]

    def empty(self) -> bool:
        return not self.q1


# Example usage
if __name__ == "__main__":
    obj = MyStack()
    obj.push(1)
    obj.push(2)
    print(obj.top())    # 2
    print(obj.pop())    # 2
    print(obj.empty())  # False
